streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableReduce.java

```diff
@@ -73,7 +73,17 @@ public void process(final K key, final Change<V> value) {
             final V oldAgg = store.get(key);
             V newAgg = oldAgg;
 
-            // first try to add the new value
+            // first remove the old value
```

Comment:
I've moved this forward to exercise an extra condition: if my thinking is correct, we shouldn't need to (potentially) initialize the aggregate first. If oldValue is non-null, then we should have a non-null aggregate already in the state, period. If we observe anything different, then the state is inconsistent with the incoming stream, and something bad has happened. Like: either we lost the state or the topology changed.

Is that right?

Clearly the prior code was more permissive, but it seems better in this case to be strict and fail fast, so we don't cause corrupted data to be sent downstream.

Reply:
I don't know if the aggregate value should be already non-null if KTableReduceProcessor.process() is called with non-null oldValue, but this logic is not expressed in the code, and I guess you are basing it on the assumption that Change instances arrive here with either newValue or oldValue null, but it doesn't seem to be enforced by Change class either, as you have explained it happens in KTableRepartitionMap, so from the point of view of KTableReduceProcessor, that happens very far is a different part of the system, and it seems a bit unsafe to rely on that behaviour in a completely unrelated class.

Just looking at this class and neighboring classes (Change) it seems that KTableReduceProcessor.process() can be called with oldAgg=null, and some Change value with both newValue and oldValue both having non-null values. You have explained that it is not possible, and with the way the code will be with this changeset, that would result in an exception, even though previously it would've worked. (As you have noticed, new behavior is more restrictive).
As far as I understand that's not possible currently because in a larger picture these classes are instantiated and connected in a certain way by a higher level API, but just looking at this piece of code that feels a bit weird.

So maybe a comment explaining why is that impossible would have been helpful. Or maybe represent a Change after it has been deserialized as a class that guarantees that only one of the newValue/oldValue is present, and make a notice that we first expect to see a newValue first always.

Reply:
I agree with @redvasily that the `KTableReduceProcessor` itself does not enforce this, and it is only enforced by the bigger scoped syntax that `KTableReduce` is part of a KTable#groupBy#reduce, in which case a `KTableRepartitionMap` is ALWAYS before the `KTableReduce`. It means that if in the future we would break this assumption, say in our optimization (really do not know if we're ever gonna do it), then this enforced check will fail.

To help @redvasily 's scenario to better be notified that a topology may have changed, I think a more general solution would be remembering the previous topology somewhere, and then after a restart checking if the topology has changed or not, than enforcing this check here. Of course that is a much larger feature to add, and that's why I'm lgtm on this PR itself, baring in mind that we may need to remove this change in the future if we ever change the way of reduce implementation that `KTableRepartitionMap` is not always used before `KTableReduce`.

Reply:
Hey @redvasily @guozhangwang ,

Sorry for the silence, I'm just coming back after paternity leave.

I may have missed the point, but I don't think this check has anything to do with whether Change has either new or old value or both. And I actually don't think that it makes any assumption about the upstream operators either.

The invariant is simply that if we observe `value.oldValue == X`, then we have previously observed `value.newValue == X`, and therefore `newAgg != null`.

Reply:
@vvcephei I see your point now. Originally I thought you are trying to resolve @redvasily 's issue of non-awared topology change, but in fact you're trying to tighten the screws here a little, and helping @redvasily 's case is like a by-product of it.

In this case, why we'd only want to do it in KTableReduce, not in KTableAggregate?

Reply:
Good point. `KTableAggregate` is slightly different, in that there is an initializer.

However, I think the same reasoning applies: if `oldValue` is non-null, then the loaded aggregate should also not be null. We have an opportunity to catch bugs by actually checking the condition.

I can add such a change to `KTableAggregate`